Tests/SonarScanner.Shim.Tests/PropertiesFileGeneratorTests.cs

```diff
@@ -724,11 +724,113 @@ public void ProjectData_Orders_AnalyzerOutPaths()
+        }
+
+        [TestMethod]
+        public void GetClosestProjectOrDefault_WhenNoProjects_ReturnsNull()
+        {
+            // Arrange & Act
+            var actual = PropertiesFileGenerator.GetSingleClosestProjectOrDefault(new FileInfo("foo"), Enumerable.Empty<ProjectData>());
+
+            // Assert
+            actual.Should().BeNull();
+        }
+
+        [TestMethod]
+        public void GetClosestProjectOrDefault_WhenNoMatch_ReturnsNull()
+        {
+            // Arrange
+            var projects = new[]
+            {
+                new ProjectData(new ProjectInfo { FullPath = "D:\\foo.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "~foo\\bar.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "C:\\foobar.csproj" }),
+            };
+
+            // Act
+            var actual = PropertiesFileGenerator.GetSingleClosestProjectOrDefault(new FileInfo("E:\\foo"), projects);
 
+            // Assert
+            actual.Should().BeNull();
+        }
+
+        [TestMethod]
+        public void GetClosestProjectOrDefault_WhenOnlyOneProjectMatchingWithSameCase_ReturnsProject()
+        {
+            // Arrange
+            var projects = new[]
+            {
+                new ProjectData(new ProjectInfo { FullPath = "foo.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "~foo\\bar.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\foo.csproj" }),
+            };
 
+            // Act
+            var actual = PropertiesFileGenerator.GetSingleClosestProjectOrDefault(new FileInfo("C:\\foo\\foo.cs"), projects);
+
+            // Assert
+            actual.Should().Be(projects[2]);
+        }
+
+        [TestMethod]
+        public void GetClosestProjectOrDefault_WhenOnlyOneProjectMatchingWithDifferentCase_ReturnsProject()
+        {
+            // Arrange
+            var projects = new[]
+            {
+                new ProjectData(new ProjectInfo { FullPath = "foo.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "~foo\\bar.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\foo.csproj" }),
+            };
+
+            // Act
+            var actual = PropertiesFileGenerator.GetSingleClosestProjectOrDefault(new FileInfo("C:\\FOO\\FOO.cs"), projects);
+
+            // Assert
+            actual.Should().Be(projects[2]);
+        }
+
+        [TestMethod]
+        public void GetClosestProjectOrDefault_WhenMultipleProjectsMatch_ReturnsProjectWithLongestMatch()
+        {
+            // Arrange
+            var projects = new[]
+            {
+                new ProjectData(new ProjectInfo { FullPath = "C:\\foo.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\bar.csproj" }),
+                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\bar\\foo.csproj" }),
```

Comment:
Minor: this test checks that a shorter match is replaced by a longer one, but not that a longer match is not replaced by a shorter one (i.e. this test would still pass if the product code was just using the last match in the list, irrespective of length).

Adding another shorter project at the end would fix this e.g.
> 
                new ProjectData(new ProjectInfo { FullPath = "C:\\foo.csproj" }),
                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\bar.csproj" }),
                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\bar\\foo.csproj" }),
                new ProjectData(new ProjectInfo { FullPath = "C:\\foo\\xxx.csproj" }),

Reply:
We also expect the product code to correctly handle a mixture of directory separator characters - different characters within the same file name/directory name, and different separators in the directory from the file (at least, I think we do).
Some of those cases will be handled by the _PathHelper_, but the _GetClosest_ method is also doing some string-content checking of its own so it would be worth using a variety path separators too.

Reply:
Regarding your 2nd comment I can't see where you see some check that could be impacted. I am working only with FileInfo and DirectoryInfo so the separator are normalized. Then I only check for length.

Anyway I'll add more tests to be sure.

Reply:
No need. My mistake - I was thinking about the implementation of the method (I spent a bit of time looking at it trying to work out where it could be going wrong).